Added HomeKit fan speed based on speed_list

nd-net opened the pull request:
## Description:


Adds support the `speed` of HomeKit fans. The commit uses `speed_list` to provide a mapping between the step-based `speed` of Home Assistant fans and the percentage-based `RotationSpeed` of HomeKit. Once [home-assistant/architecture#27](https://github.com/home-assistant/architecture/issues/27) changes the architecture, the `speed` should be retrieved via the new architecture.

**Related issue (if applicable):** updates #14351

**Pull request in home-assistant.github.io with documentation (if applicable):** home-assistant/home-assistant.io#8044

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)
  - [x] Tests have been added to verify that the new code works.

nd-net committed:
Added HomeKit fan speed based on speed_list

cdce8p commented:
I closed a similar PR a while back #14942 since some platforms use their own speed lists and all options would lead to problems down the road.

However I need to acknowledge that this is a highly requested feature. Since home-assistant/architecture#27 hasn't gone anywhere and isn't likely to change in the future, we might as well add rudimentary support now.

I would like to suggest limiting the support to the four base modes for now: `off`, `low`, `medium` and `high`. Also in the light of https://github.com/home-assistant/architecture/issues/127 and more possible changes to the fan component. Support for different once could be added later.

I'll go through your code in a few minutes to give you a first review.

cdce8p commented on homeassistant/components/homekit/type_fans.py:
```diff
@@ -41,19 +46,29 @@ def __init__(self, *args):
         if features & SUPPORT_OSCILLATE:
             chars.append(CHAR_SWING_MODE)
+        if features & SUPPORT_SET_SPEED:
+            chars.append(CHAR_ROTATION_SPEED)
```
Before adding the `char`, we would need to make sure we support the specific speed list.
Maybe extract that check into a new method, something along the lines of:
```py
def check_speed_list(self):
  speed_list = self.hass.states.get(self.entity_id).attributes.get(ATTR_SPEED_LIST)
  if speed_list == [SPEED_OFF, SPEED_LOW, SPEED_MEDIUM, SPEED_HIGH]:
    return True
  return False
```

```py
if features & SUPPORT_SET_SPEED and self.check_speed_list():
    chars.append(CHAR_ROTATION_SPEED)
```

cdce8p replied:
As an alternative `check_speed_list` could return an integer we use for internal mapping: `speed_list_type`.

nd-net replied:
As long as speed_list contains arbitrary values, I cannot think of a way of using and integer. **Only if** we are restricting the list to the base modes I imagine that I can use a constant indicating the possibilities _speed list with off, low, high_ and _speed list with off, low, medium, high_, but this would be very brittle and break even if something trivial like a fourth speed suggested in home-assistant/architecture#127 gets introduced.

I think a better idea something that contains a mapping between Home Assistant speed values and percentages; the percentages must allow for ranges since a fan with `off`, `low`, `high` should behave like this to be intuitive (IMHO):

- 0%-33% should map to `off`
- `off` should be shown as 0% (otherwise the user as the strange experience that the fan is off, but still shows a non-zero value)
- 34%-66% should map to `low`
- `low` should be shown as 50%
- 67%-100% should map to `high`
- `high` should be shown as 100%

The bounds of the ranges can be set either implicitly or explicitly. If everything is implicit, then this would be equivalent to storing `speed_list`, just as I did in this PR.

I think the best way to do this is to use list/set containing `namedtuple("PercentageMapping", "speed_value percentage_target percentage_start percentage_stop")` for the `speed_list_mapping`.

Another way is to use a dict: `speed_value_to_percentage` and a sorted list containing tuples(`percentage_stop`, `speed_value`), but in my opinion the first option is easier to understand, read, and debug.

cdce8p replied:
Please ignore the architecture issues for now.

As for the mapping, I agree. Sonething like:
Speed -> Target, Start, Stop
does make sense.

cdce8p commented on homeassistant/components/homekit/type_fans.py:
```diff
@@ -83,6 +98,56 @@ def set_oscillating(self, value):
                   ATTR_OSCILLATING: oscillating}
         self.call_service(DOMAIN, SERVICE_OSCILLATE, params, oscillating)
 
+    def set_speed_list(self, speed_list):
```
We don't need to save the complete speed list inside the accessory. After the check the `speed_list_type` should be enough for the conversion later.

nd-net replied:
Still new to Home Assistant, I thought that maybe this should support later alterations of speed_list via the Customization UI; but I do agree that assuming that the list stays the same makes the code simpler and easier to understand.

cdce8p commented on homeassistant/components/homekit/type_fans.py:
```diff
@@ -83,6 +98,56 @@ def set_oscillating(self, value):
+                SPEED_OFF, speed_list[0])
+        self.speed_list = speed_list
+
+    def speed_from_percentage(self, percentage):
```
`speed_from_percentage` and `percentage_from_speed` should be moved to `homekit/util.py`.
What do you think of:
 - `speed_to_homekit`
 - `speed_to_states`

That's the naming schema we used for temperature values.

nd-net replied:
Okay, good catch that this should be part of a utility. However, if we do this, maybe we should think ahead: 
home-assistant/architecture#27 mentions that other secondary platforms most often expect percentages, therefore I think it would also fit into either `/util/fan_speed.py` or into `/components/fan/util.py`. When doing this, then `*_homekit` would be a bit misleading.

I think of a small class that contains the mapping defined in https://github.com/home-assistant/home-assistant/pull/19767#discussion_r245417379 and that contains has the following methods: `__init__(speed_list: List[Any])`, `as_speed(percentage: int) -> Any` and `as_percent(speed: Any) -> int`. Alternatively, there can be a free-floating function that returns the mapping data structure and 2 such functions for the conversion, but on the other hand this is what classes are for - to group data and associated functions.

cdce8p commented on homeassistant/components/homekit/type_fans.py:
```diff
@@ -83,6 +98,56 @@ def set_oscillating(self, value):
+    def speed_from_percentage(self, percentage):
+        """Convert a value from 0-100 to a value in the speed list."""
+        if not self.speed_list:
+            return None
```
If we pass the `speed_list_type` to those methods, we should be able to do direct mapping. No need for other checks and calculations. The mapping can be though a dictionary.

nd-net replied:
Agreed

cdce8p commented on homeassistant/components/homekit/type_fans.py:
```diff
@@ -83,6 +98,56 @@ def set_oscillating(self, value):
+        self._flag[CHAR_ROTATION_SPEED] = True
+        speed = self.speed_from_percentage(value)
+        params = {ATTR_ENTITY_ID: self.entity_id,
+                  ATTR_SPEED: speed}
```
```py
ATTR_SPEED: speed_to_states(speed, self.speed_list_type)

nd-net replied:
Now using generic `self.speed_mapping.speed_to_states(value)` to map the attribute to a percentage from HA states

cdce8p commented on homeassistant/components/homekit/type_fans.py:
```diff
@@ -104,6 +169,18 @@ def update_state(self, new_state):
 
+        # Handle Speed
+        if self.char_speed is not None:
+            self.set_speed_list(new_state.attributes.get(ATTR_SPEED_LIST))
```
Getting a new `speed_list` shouldn't be necessary, since it shouldn't change.

cdce8p commented on homeassistant/components/homekit/type_fans.py:
```diff
@@ -26,11 +28,14 @@ class Fan(HomeAccessory):
     Currently supports: state, speed, oscillate, direction.
     """
 
+    speed_list = []
```
Isn't necessary if we use `self.speed_list_type`.

cdce8p reviewed (changes requested):
Some first comments. I haven't taken a look at the tests yet. Will do that once the basic structure is finished.

cdce8p commented:
I though as much. All the issues you mentioned are the reason why we haven't added the speed support yet.

As for the speed modes, I would like to start with the base ones and expand from there. My suggestion with some kind of `speed_list_type` combined with direct mapping would allow us to cover most standard plaforms.

However for the mqtt and template ones we would need to expand that concept. I was originally thinking of adding a config option that contains the mapping dict for that. If that's ok for you, we should first cover the basics and then expand it in a subsequent PR.

marchingphoenix commented:
For `google_assistant` i was directed to only support the base fan speeds. Not sure if that same directive holds true here too.

nd-net commented:
Okay, thank you @marchingphoenix; I just did look at your code from `google_assistant` as well. From what I can see Google Assistant uses text-based speed settings as well and you did map Home Assistant and Google Assistant.

If I should go the same route, I would need all of your input on how a mapping should work for percentages.
The extremes are easy: `off` -> 0%; `high` -> 100%. For the other possibilities I would need those `speed_list_types`:

* 2-speed fan with `low`: `low` maps range of 34%-66% with `speed_to_homekit` = 50%
* (2-speed fan with `medium`: `medium ` maps range of 34%-66% with `speed_to_homekit` = 50%)
* 3-speed fan: `low ` maps range of 26%-50% with `speed_to_homekit` = 33%, `medium ` maps range of 51%-75% with `speed_to_homekit` = 66%

Is this something that I should implement?

cdce8p commented:
@nd-net Regarding the specific mapping, I think those are details we can decide later once the structure is finalized. Ultimately it will depend on the speed list type as you already know.

nd-net committed:
Moved speed_list <=> HomeKit conversion to util

Incorporated feedback from Code Review

* Moved mapping between speed_list and values expected by HomeKit into util.py
* Updated tests
* speed_list gets checked at initialization-time

The updated HomeKit Fan still supports arbitrary values for the speed_settings; however, it expects that the first element is equivalent to `off`.

nd-net commented:
@cdce8p Implemented the requested changes; the fan speed can be arbitrary as mentioned in https://github.com/home-assistant/home-assistant/pull/19767#issuecomment-451573319 .

cdce8p commented:
@nd-net Sorry for the delay. I don't have much spare time at the moment. Hope to get to it this weekend at the latest.

cdce8p commented on homeassistant/components/homekit/util.py:
```diff
@@ -110,6 +113,59 @@ def validate_media_player_features(state, feature_list):
+            stop = (index + 1) * 100 / list_size
+            # the first and last element should go until -+ infinity
+            if index == 0:
+                start = -inf
```
The minimum HomeKit speed value is `0`. Based on your implementation below, the following should work
```py
start = 0
```

For stop it's similar. The max value is `100`, so `stop = 101` should be enough.

cdce8p commented on homeassistant/components/homekit/util.py:
```diff
@@ -110,6 +113,59 @@ def validate_media_player_features(state, feature_list):
+            # * index = len(speed_list) - 1 => 100 %
+            # * all other indices are equally distributed
+            target = index * 100 / (list_size - 1)
+            start = index * 100 / list_size
```
```suggestion
            start = index * 100 / list_size if index != 0 else 0
```
The if clause below can be deleted.

cdce8p commented on homeassistant/components/homekit/util.py:
```diff
@@ -110,6 +113,59 @@ def validate_media_player_features(state, feature_list):
+            # * all other indices are equally distributed
+            target = index * 100 / (list_size - 1)
+            start = index * 100 / list_size
+            stop = (index + 1) * 100 / list_size
```
```suggestion
            stop = (index + 1) * 100 / list_size if index != list_size - 1 else 101
```

The if clause below can be deleted.

cdce8p commented on homeassistant/components/homekit/util.py:
```diff
@@ -110,6 +113,59 @@ def validate_media_player_features(state, feature_list):
+                            "%s as its first element. "
+                            "Assuming that %s is equivalent to 'off'.",
+                            speed_list, fan.SPEED_OFF, speed_list[0])
+        speed_ranges = {}
```
If we use and `OrderedDict`, we should be able to assum later that the first item is the min speed setting. See `speed_to_states` below.

cdce8p commented on homeassistant/components/homekit/util.py:
```diff
@@ -110,6 +113,59 @@ def validate_media_player_features(state, feature_list):
+                return state
+        # if nobody tampered with self.speed_ranges
+        # this happens only for +/-inf
+        raise KeyError(speed)
```
With the removal of `inf`. This can probably be deleted as well.
Although this probably doesn't happen, with an `OrderedDict` we could return a default value
```py
return speed_ranges.keys()[0]

nd-net replied:
Unfortunately, keys() is a view which cannot be indexed. However, the case still should not happen, so I keep raising a KeyError

cdce8p replied:
My fault, add `list(speed_ranges.keys())[0]`

cdce8p commented on tests/components/homekit/test_type_fans.py:
```diff
@@ -155,3 +158,51 @@ def cls():
+    await hass.async_block_till_done()
+    acc = cls.fan(hass, hk_driver, 'Fan', entity_id, 2, None)
+
+    assert acc.char_speed is None
```
Add this to `test_fan_basic` instead.

cdce8p commented on tests/components/homekit/test_type_fans.py:
```diff
@@ -155,3 +158,51 @@ def cls():
+    acc = cls.fan(hass, hk_driver, 'Fan', entity_id, 2, None)
+
+    assert acc.char_speed.value == 0
+
```
Empty line can probably be deleted

cdce8p commented on tests/components/homekit/test_type_fans.py:
```diff
@@ -155,3 +158,51 @@ def cls():
+
+    await hass.async_add_job(acc.run)
+
+    speed_mapping = acc.speed_mapping
```
I think this line isn't necessary. Just to ommit `acc.` below?

cdce8p commented on tests/components/homekit/test_util.py:
```diff
@@ -144,3 +145,62 @@ def test_density_to_air_quality():
         HOMEKIT_NOTIFY_ID
+
+
+def test_homekit_speed_mapping_creates_correct_speed_ranges():
```
```suggestion
def test_homekit_speed_mapping():
```

cdce8p reviewed (changes requested):
@nd-net I finally got to review your changes. Sorry again for the long delay. All in all your changes look pretty good. At the beginning I wasn't so convinced that this would work, but I'm starting to believe it will. Great job so far!

I've added some minor comments.

nd-net commented on homeassistant/components/homekit/util.py:
```diff
@@ -110,6 +113,59 @@ def validate_media_player_features(state, feature_list):
 
 
+SpeedRange = namedtuple('SpeedRange', 'start target stop')
+SpeedRange.__doc__ += """: a mapping of descrete Home Assistant speed \
```
Okay, but I still need to reference [__doc__](https://docs.python.org/3.5/library/collections.html#collections.somenamedtuple._fields) field of named tuple. Other instances of `namedtuple` in home-assistant are without comment, but I would really like to document what is inclusive and what not.

cdce8p replied:
You could, but unless the linter complaints, I don't think that it's necessary. A multiline code comment is probably enough explanation.

nd-net commented on homeassistant/components/homekit/type_fans.py:
```diff
@@ -83,6 +102,15 @@ def set_oscillating(self, value):
+        self._flag[CHAR_ROTATION_SPEED] = True
+        speed = self.speed_mapping.speed_to_states(value)
+        params = {ATTR_ENTITY_ID: self.entity_id,
+                  ATTR_SPEED: speed}
```
If I am doing this, then I don't have a value element for the `call_service` call below. As far as I can see this is nothing to worry about, but I just wanted you to know.

cdce8p replied:
You're right. Please ignore my comment.

nd-net committed:
Simplified speed mapping

Incorporated feedback from Code Review

Using OrderedDict for speed mapping: stop not needed anymore

Removed unnecessary code

cdce8p commented on homeassistant/components/homekit/util.py:
```diff
@@ -110,6 +112,51 @@ def validate_media_player_features(state, feature_list):
+
+    def __init__(self, speed_list):
+        """Initialize a new SpeedMapping object."""
+        if not speed_list or speed_list[0] != fan.SPEED_OFF:
```
The case `speed_list is None` shouln't happen. I was thinking if we really need it here if we do it should probably be checked in `type_fans` first, so we don't even add the char. However since it really shouldnkt happen, it's probably save to delete.

cdce8p reviewed (changes requested):
Some last suggestions

nd-net commented on homeassistant/components/homekit/util.py:
```diff
@@ -110,6 +112,51 @@ def validate_media_player_features(state, feature_list):
 
 
+SpeedRange = namedtuple('SpeedRange', ('start', 'target'))
+SpeedRange.__doc__ += """: Maps Home Assistant speed \
```
The default __doc__ of a named tuple is _name_(_arg_, ...) - I will remove the colon, but the space should still be present to make sure it's not _SpeedRange(start, target)Maps_.

nd-net committed:
Removed unnecessary code

nd-net committed:
Removed check for empty speed list

cdce8p commented on homeassistant/components/homekit/util.py:
```diff
@@ -110,6 +112,51 @@ def validate_media_player_features(state, feature_list):
+            target = index * 100 / (list_size - 1)
+            start = index * 100 / list_size
+            speed_ranges[speed] = SpeedRange(start, target)
+        self.speed_ranges = speed_ranges
```
What do you think here?

nd-net committed:
Fixed code-review issues

cdce8p committed:
Add debounce to set_speed

cdce8p commented:
During the last check I noticed that `debounce` was missing for `set_speed`. Since everything else is done, I've just added it. Can you update the doc PR according to the comments? https://github.com/home-assistant/home-assistant.io/pull/8044
Thanks for the great work you did here! :1st_place_medal:
I'll merge it, once that tests pass.
